pkg/webhook/preflight/preflight.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2025 Nutanix. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package preflight
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type Cause struct {
+	Message string
+	Field   string
+}
+
+type CheckResult struct {
+	Name string
+
+	Allowed bool
+	Error   bool
+
+	Causes   []Cause
+	Warnings []string
+}
+
+type Check = func(ctx context.Context) CheckResult
+
+type Checker interface {
+	// Init decides which of its checks should run for the cluster. It then initializes the checks
+	// with common dependencies, such as an infrastructure client. Finally, it returns the initialized checks,
+	// ready to be run.
+	//
+	// Init can store the client and cluster, but not the context, because each check will accept its own context.
+	Init(ctx context.Context, client ctrlclient.Client, cluster *clusterv1.Cluster) []Check
+}
+
+type WebhookHandler struct {
+	client   ctrlclient.Client
+	decoder  admission.Decoder
+	checkers []Checker
+}
+
+func New(client ctrlclient.Client, decoder admission.Decoder, checkers ...Checker) *WebhookHandler {
+	h := &WebhookHandler{
+		client:   client,
+		decoder:  decoder,
+		checkers: checkers,
+	}
+	return h
+}
+
+func (h *WebhookHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if req.Operation == admissionv1.Delete {
+		return admission.Allowed("")
+	}
+
+	cluster := &clusterv1.Cluster{}
+	err := h.decoder.Decode(req, cluster)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Checks run only for ClusterClass-based clusters.
+	if cluster.Spec.Topology == nil {
+		return admission.Allowed("")
+	}
+
+	resp := admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{
+			Allowed: true,
+			Result: &metav1.Status{
+				Details: &metav1.StatusDetails{},
+			},
+		},
+	}
+
+	// Collect all checks in parallel.
+	checkerWG := &sync.WaitGroup{}
+	resultCh := make(chan CheckResult)
+	for _, checker := range h.checkers {
+		checkerWG.Add(1)
+
+		go func(ctx context.Context, checker Checker, resultCh chan CheckResult) {
+			// Initialize the checker.
+			checks := checker.Init(ctx, h.client, cluster)
+
+			// Run its checks in parallel.
+			checksWG := &sync.WaitGroup{}
+			for _, check := range checks {
+				checksWG.Add(1)
+				go func(ctx context.Context, check Check, resultCh chan CheckResult) {
+					result := check(ctx)
+					resultCh <- result
+					checksWG.Done()
+				}(ctx, check, resultCh)
+			}
+			checksWG.Wait()
+
+			checkerWG.Done()
+		}(ctx, checker, resultCh)
+	}
+
+	// Close the channel when all checkers are done.
+	go func(wg *sync.WaitGroup, resultCh chan CheckResult) {
+		wg.Wait()
+		close(resultCh)
+	}(checkerWG, resultCh)
```

Comment:
can the function return before this go routing closes the channel?

what happens with slow checks?

Reply:
I replaced this code when I switched approaches, but your question is still valid. And it's a good question :smile: 

> still would like to know how slow checks end up getting propogated up to the user!

The short answer is: they don't.

The webhook configuration has a configurable timeout, up to 30 seconds. The API server abandons the request when the time is up. No results are returned at all, and the user merely sees a "context deadline exceeded" error.

We want to return the results of some checks to the user, even if we can't complete all the checks in time. To do that, the framework will need to (a) impose its own timeout (and it must be shorter than the one configured for the webhook), (b) pre-empt checks that fail to return within the timeout.

Reply:
I think it's important to return partial results instead of simply timing out. I'll make sure we address this in a future PR. How does that sound?

Reply:
I filed https://jira.nutanix.com/browse/NCN-107885